File: mobile_app/lib/core/routes/role_routing.dart
import '../../features/auth/models/auth_user.dart';
import 'app_routes.dart';

class RoleRouting {
  RoleRouting._();

  static String? dashboardForRole(String? role) {
    switch (role?.toLowerCase()) {
      case 'admin':
        return AppRoutes.adminDashboard;
      case 'specialist':
        return AppRoutes.specialistDashboard;
      case 'parent':
        return AppRoutes.parentDashboard;
      default:
        return null;
    }
  }

  static String? normalizeSpecialistVerificationStatus(String? status) {
    final normalized = status?.trim().toLowerCase();
    if (normalized == 'approved' ||
        normalized == 'rejected' ||
        normalized == 'pending') {
      return normalized;
    }
    return null;
  }

  static String? getSpecialistVerificationStatus(AuthUser? user) {
    if (user?.role?.toLowerCase() != 'specialist') {
      return null;
    }

    return normalizeSpecialistVerificationStatus(user?.verificationStatus) ??
        'pending';
  }

  static bool isApprovedSpecialist(AuthUser? user) {
    return getSpecialistVerificationStatus(user) == 'approved';
  }

  /// Role home that also routes unapproved specialists to verification screens.
  static String? homeForUser(AuthUser? user) {
    final role = user?.role?.toLowerCase();
    if (role != 'specialist') {
      return dashboardForRole(role);
    }

    final status = getSpecialistVerificationStatus(user);
    if (status == 'approved') {
      return AppRoutes.specialistDashboard;
    }
    if (status == 'rejected') {
      return AppRoutes.specialistVerificationRejected;
    }
    return AppRoutes.specialistVerificationPending;
  }

  static bool isAdminRoute(String path) => path.startsWith('/dashboard/admin');

  static bool isSpecialistRoute(String path) =>
      path.startsWith('/dashboard/specialist');

  static bool isParentRoute(String path) => path.startsWith('/dashboard/parent');

  static bool isSpecialistVerificationRoute(String path) {
    return path == AppRoutes.specialistVerificationPending ||
        path == AppRoutes.specialistVerificationRejected ||
        path.startsWith('/specialist-verification/');
  }

  static bool isProtectedDashboardRoute(String path) {
    return isAdminRoute(path) || isSpecialistRoute(path) || isParentRoute(path);
  }

  static bool canAccessRoute(AuthUser? user, String path) {
    final role = user?.role?.toLowerCase();

    if (isAdminRoute(path)) {
      return role == 'admin';
    }
    if (isSpecialistRoute(path)) {
      return role == 'specialist' && isApprovedSpecialist(user);
    }
    if (isParentRoute(path)) {
      return role == 'parent';
    }
    if (isSpecialistVerificationRoute(path)) {
      return role == 'specialist';
    }
    return true;
  }
}

File: frontend_web/src/routes/roleRouting.js
const ROLE_DASHBOARD_MAP = {
  admin: "/dashboard/admin",
  specialist: "/dashboard/specialist",
  parent: "/dashboard/parent",
};

export const SPECIALIST_VERIFICATION_PENDING_PATH =
  "/specialist-verification/pending";
export const SPECIALIST_VERIFICATION_REJECTED_PATH =
  "/specialist-verification/rejected";

function normalizeRole(role) {
  return typeof role === "string" ? role.trim().toLowerCase() : "";
}

export function normalizeSpecialistVerificationStatus(status) {
  const normalized = typeof status === "string" ? status.trim().toLowerCase() : "";
  if (normalized === "approved" || normalized === "rejected" || normalized === "pending") {
    return normalized;
  }
  return null;
}

export function getSpecialistVerificationStatus(user) {
  if (normalizeRole(user?.role) !== "specialist") {
    return null;
  }

  return (
    normalizeSpecialistVerificationStatus(
      user?.verification_status ?? user?.verificationStatus,
    ) || "pending"
  );
}

export function isApprovedSpecialist(user) {
  return getSpecialistVerificationStatus(user) === "approved";
}

export function dashboardForRole(role) {
  return ROLE_DASHBOARD_MAP[normalizeRole(role)] || null;
}

/**
 * Role home that also routes unapproved specialists to verification screens.
 */
export function homeForUser(user) {
  const role = normalizeRole(user?.role);
  if (role !== "specialist") {
    return dashboardForRole(role);
  }

  const status = getSpecialistVerificationStatus(user);
  if (status === "approved") {
    return ROLE_DASHBOARD_MAP.specialist;
  }
  if (status === "rejected") {
    return SPECIALIST_VERIFICATION_REJECTED_PATH;
  }
  return SPECIALIST_VERIFICATION_PENDING_PATH;
}

export function isSpecialistVerificationRoute(pathname) {
  return (
    pathname === SPECIALIST_VERIFICATION_PENDING_PATH
    || pathname === SPECIALIST_VERIFICATION_REJECTED_PATH
    || pathname.startsWith("/specialist-verification/")
  );
}

export function isProtectedDashboardRoute(pathname) {
  return pathname === "/dashboard" || pathname.startsWith("/dashboard/");
}

export function canAccessRoute(role, pathname, user = null) {
  const normalizedRole = normalizeRole(role ?? user?.role);

  if (pathname === "/dashboard") {
    return Boolean(homeForUser(user || { role: normalizedRole }));
  }

  if (pathname.startsWith("/dashboard/admin")) {
    return normalizedRole === "admin";
  }

  if (pathname.startsWith("/dashboard/specialist")) {
    return normalizedRole === "specialist" && isApprovedSpecialist(user || { role: normalizedRole });
  }

  if (pathname.startsWith("/dashboard/parent")) {
    return normalizedRole === "parent";
  }

  if (isSpecialistVerificationRoute(pathname)) {
    return normalizedRole === "specialist";
  }

  return !isProtectedDashboardRoute(pathname);
}

export function getRoleLabel(role) {
  const normalizedRole = normalizeRole(role);

  if (normalizedRole === "admin") {
    return "Admin";
  }

  if (normalizedRole === "specialist") {
    return "Specialist";
  }

  if (normalizedRole === "parent") {
    return "Parent";
  }

  return "User";
}
